### PR TITLE
feat: fixes for lead form

### PR DESCRIPTION
**Pull request by Averethel**
References https://autoricardo.atlassian.net/browse/DM-438

## Changes

- exposes `FormLabel` component so that labels can be used outside of `FormControl` (eg. for checkbox groups)
- exposes `Text` component to render styleable text
- adds `type` property to the `Button` so that we can render `submit`
- makes `TextArea` component full width so it behaves the same way as `Input`
- fixes `FormLabel` text color to `gray.900`

## How to test

New components are exposed in storybook

### CODE REVIEW

File: src/components/button/index.tsx [hunk 1]
```diff
@@ -1,18 +1,30 @@
...
+  onClick?: (event: MouseEvent<HTMLElement>) => void;
+};
+
+type Props = ButtonProps | SubmitProps;
 
 const Button: FC<Props> = ({
   variant = 'primary',
   size = 'lg',
   isDisabled = false,
+  type = 'button',
```
Comment (talamcol) [inline src/components/button/index.tsx:27]: can you confirm this behavior with the types:

When I do not specify a type (which defaults to "button") it does not complain if I do not pass an `onClick` handler. If I however specify the type explicitly to `button` it does complain.

File: src/components/button/index.tsx [hunk 2]
```diff
@@ -1,18 +1,30 @@
...
   size?: 'md' | 'lg';
   isDisabled?: boolean;
-  onClick: (event: MouseEvent<HTMLElement>) => void;
   children: ReactNode;
-}
+};
+
+type ButtonProps = SharedProps & {
+  type?: 'button';
+  onClick: (event: MouseEvent<HTMLElement>) => void;
```
Comment (talamcol) [inline src/components/button/index.tsx:13]: it's not part of this PR but could we use the type from ChakraButtonProps for the onClick? It feels wrong to type it again since it should be the native HTML event type. Maybe `Pick<>` could help here

Bildschirmfoto 2022-08-08 um 09 04 55

File: src/components/text/index.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import React, { FC } from 'react';
+import { Text as ChakraText, TextProps } from '@chakra-ui/react';
+
+type MarginProps = 'm' | 'mt' | 'mb' | 'mr' | 'ml' | 'my' | 'mx';
```
Comment (talamcol) [inline src/components/text/index.tsx:4]: could we use `SpaceProps` or do you want to avoid that devs pass marginX instead of mx?

File: src/themes/components/textarea.ts [hunk 4]
```diff
@@ -13,6 +13,7 @@ const baseStyle: SystemStyleObject = {
   appearance: 'none',
   transitionProperty: 'common',
   transitionDuration: 'normal',
+  width: '100%',
```
Comment (talamcol) [inline src/themes/components/textarea.ts:16]:
```suggestion
  width: 'full',
```

**Review by talamcol — COMMENTED** [review]
it looks good to me

File: src/components/button/index.tsx [hunk 5]
```diff
@@ -1,18 +1,30 @@
...
+};
+
+type SubmitProps = SharedProps & {
```
Comment (lkappeler) [inline src/components/button/index.tsx:16]: Why do you need to split the props and cannot just `type?: 'button | 'submit'`?

File: src/components/formLabel/index.tsx [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import React, { FC, PropsWithChildren } from 'react';
+import { FormLabel as ChakraFormLabel } from '@chakra-ui/react';
+
+const FormLabel: FC<PropsWithChildren> = ({ children }) => (
```
Comment (lkappeler) [inline src/components/formLabel/index.tsx:4]: Should we update the `FormControl` component to consume the customised `FormLabel` (I see we don't customise it yet but we might do it in the future and this will avoid diverging the styles...)

File: src/components/text/index.tsx [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+type PaddingProps = 'p' | 'pt' | 'pb' | 'pr' | 'pl' | 'py' | 'px';
+
+type Props = Pick<
```
Comment (lkappeler) [inline src/components/text/index.tsx:7]: Do we really need to expose all these combinations for a text? This allows for very strange combinations I doubt we will ever use in any layout. For example something like this

image

File: src/components/button/index.tsx (hunk 1 above)
Comment (lkappeler) [reply src/components/button/index.tsx:27]: ran into the same issue, made a quick PR: [link the ticket here]

Comment (Averethel) [reply src/components/button/index.tsx:27]: That was intended. `onClick` was mandatory for `button` but we wanted to make it optional for `submit`

File: src/components/button/index.tsx (hunk 2 above)
Comment (Averethel) [reply src/components/button/index.tsx:13]: Looks like it's taken care of by Lorenzo

File: src/components/text/index.tsx (hunk 3 above)
Comment (Averethel) [reply src/components/text/index.tsx:4]: Yeah, I wanted to harmonise the syntax we use here.

File: src/components/button/index.tsx (hunk 5 above)
Comment (Averethel) [reply src/components/button/index.tsx:16]: Because `onClick` should be mandatory for the `button`, but optional for `submit`

File: src/components/text/index.tsx (hunk 7 above)
Comment (Averethel) [reply src/components/text/index.tsx:7]: I went for a sensible minimum that gives the flexibility to implement different styles of texts without the need to reach for custom components. What would you omit here?

As the background - I was prompted to add `Text` by [privacy policy and ToC disclaimer in the lead form](https://www.figma.com/file/f0LWQupJueW6KHBQujVGe4/Lead-Form-Page-(Working-Title)?node-id=830%3A71834)

Comment (lkappeler) [reply src/components/text/index.tsx:7]: I would check back with the designer about what they intend to use/have them create a text component. Alternatively, I would expose what we need and gradually extend it when we need it for other designs while letting the designer know that they are creating a new text variation in their design. My fear is that when being open we will create a lot of small differences between the layouts (line heights, passing etc) so I d rather only provide whats intentionally scoped as a text component

File: src/components/formLabel/index.tsx (hunk 6 above)
Comment (Averethel) [reply src/components/formLabel/index.tsx:4]: After checking - we don't want it after all. They serve different purposes. You would use `FormControl` with a form field, whereas you'd use `FormLabel` only in the context of the field group. If we would want to use `FormLabel` here it would require us to allow other usages of our component (and extend the props as a consequence) - I would like to avoid that. Styling-wise they use the same theme so that will ensure a consistent look and feel.

File: src/components/button/index.tsx (hunk 1 above)
Comment (talamcol) [reply src/components/button/index.tsx:32]: What I mean is that since type=button is the default, the following should raise a TS error but it does not: `<Button>My button</Button>`. If I do this: `<Button type="button">My button</Button>` it does complain. But the two versions are identical

File: src/components/text/index.tsx (hunk 7 above)
Comment (Averethel) [reply src/components/text/index.tsx:7]: I see where this comes from, but I think there's merit to keeping it as is. This is quite a minimal set of props related to text formatting. What I can think of is only exposing `textStyle` and accepting values defined in `typography`. Do you think that's better?

I want to avoid explicitly creating a component/variant for every text style since that would add a significant overhead on our end...

File: src/components/formLabel/index.tsx (hunk 6 above)
Comment (lkappeler) [reply src/components/formLabel/index.tsx:4]: I can't follow. In the usage example, they show to use FormLabel inside the form control 
https://chakra-ui.com/docs/components/form-control/usage#usage so why would I use the FormLabel different than the chakra example?

Comment (Averethel) [reply src/components/formLabel/index.tsx:4]: Second example there shows closer what I mean

File: src/components/button/index.tsx (hunk 1 above)
Comment (Averethel) [reply src/components/button/index.tsx:32]: Hmm, let me see what I can do about that, that wasn't intended. It could be that the `type` needs to be mandatory for this to work as expected

File: src/components/text/index.tsx (hunk 7 above)
Comment (Averethel) [reply src/components/text/index.tsx:7]: as discussed - we drop the `margin` and `padding` (since it can be controlled by the layout), only use `textStyle`, but keep the `color` as is

File: src/themes/components/textarea.ts (hunk 4 above)
Comment (Averethel) [reply src/themes/components/textarea.ts]: also done for `input`

File: src/components/button/index.tsx (hunk 1 above)
Comment (Averethel) [reply src/components/button/index.tsx]: I manage to achieve that with dependent types.
Screenshot 2022-08-10 at 09 51 22

The error we have when explicitly assigning `type` to `button` is less helpful due to how types need to be structured 
Screenshot 2022-08-10 at 09 54 10
(Basically, the `type` needs to be mandatory to switch the `onClick` enforcing on and off, otherwise `undefined` will match both clauses and we can't achieve it). 
It looks like a good compromise to me, but let me know what you think

Comment (talamcol) [reply src/components/button/index.tsx]: sounds good to me 👍

**Review by talamcol — APPROVED** [review]
looks neat 🔥 thank you for the changes

File: src/components/button/index.tsx [hunk 8]
```diff
@@ -1,27 +1,40 @@
...
+import React, { FC, MouseEventHandler } from 'react';
 import {
   Button as ChakraButton,
   ButtonProps as ChakraButtonProps,
 } from '@chakra-ui/react';
 
-export type ButtonProps = {
+type OnClick = MouseEventHandler<HTMLButtonElement>;
+
+type WithOnClick<T extends 'submit' | 'button'> = T extends 'submit'
```
Comment (lkappeler) [inline src/components/button/index.tsx:9]: Optional: You could use the original definition instead of redefining the onClick with something like 
```
type WithOnClick<T extends 'submit' | 'button'> = Omit<ButtonProps, "onClick"> & T extends 'submit'
  ? { onClick?: ButtonProps['onClick'] }
  : { onClick: ButtonProps['onClick'] };
```

**Review by lkappeler — APPROVED** [review]
Nice much simpler with the text now

File: src/components/button/index.tsx (hunk 8 above)
Comment (Averethel) [reply src/components/button/index.tsx:9]: No, because the `onClick` is optional on the `ButtonProps` so it would result in `MouseEventHandler | undefined`. I used the type that chakra uses in the re-definition

Comment (Averethel) [reply src/components/button/index.tsx:9]: Actually `Exclude` exists so I can exclude `undefined` from the union 🎉
